instrumentation/racecar/test/opentelemetry/instrumentation/racecar_test.rb

```diff
@@ -0,0 +1,229 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+require 'securerandom'
+
+require 'racecar/cli'
+require_relative '../../../lib/opentelemetry/instrumentation/racecar'
+
+describe OpenTelemetry::Instrumentation::Racecar do
+  let(:instrumentation) { OpenTelemetry::Instrumentation::Racecar::Instrumentation.instance }
+  let(:exporter) { EXPORTER }
+  let(:spans) { exporter.finished_spans }
+  let(:host) { ENV.fetch('TEST_KAFKA_HOST') { '127.0.0.1' } }
+  let(:port) { (ENV.fetch('TEST_KAFKA_PORT') { 29_092 }) }
+
+  def tracer
+    OpenTelemetry.tracer_provider.tracer('test-tracer')
+  end
+
+  def produce(messages)
+    config = { "bootstrap.servers": "#{host}:#{port}" }
+    producer = Rdkafka::Config.new(config).producer
+    producer.delivery_callback = ->(_) {}
+
+    producer_messages.map do |msg|
+      tracer.in_span("#{msg[:topic]} send", kind: :producer) do
+        msg[:headers] ||= {}
+        OpenTelemetry.propagation.inject(msg[:headers])
+        producer.produce(**msg)
+      end
+    end.each(&:wait)
+
+    producer.close
+  end
+
+  let(:racecar) do
+    Racecar.config.brokers = ["#{host}:#{port}"]
+    Racecar.config.pause_timeout = 0 # fail fast and exit
+    Racecar.config.load_consumer_class(consumer_class)
+    Racecar::Runner.new(consumer_class.new, config: Racecar.config, logger: Logger.new(STDOUT), instrumenter: Racecar.instrumenter)
+  end
+
+  def run_racecar(racecar)
+    Thread.new do
+      Thread.current[:racecar] = racecar
+      racecar.run
+    rescue RuntimeError => e
+      raise e unless e.message == 'oops'
+    end
+  end
+
+  def stop_racecar(thread)
+    thread[:racecar].stop
+    thread.join(60)
+  end
+
+  def wait_for_messages_seen_by_consumer(count)
+    Timeout.timeout(60) do
+      sleep 0.1 until consumer_class.messages_seen.size >= count
+    end
```

Comment:
@open-telemetry/ruby-contrib-approvers, we should keep an eye on this to make sure the tests so we can avoid non-deterministic tests. This may be a good use case for a CountDownLatch.

Reply:
Reading this again, I'd like to tidy this up a little too. I'll introduce CountDownLatch and remove some of the duplication.